kore/src/Kore/Rewrite/Axiom/Matcher.hs

```diff
@@ -81,696 +68,401 @@ import Kore.Internal.Predicate (
-    c1 <= c2 = termPriority c1 <= termPriority c2
+type Value normalized =
+    Builtin.Value normalized (TermLike RewritingVariableName)
 
-findClass :: Constraint variable -> TermLikeClass
-findClass (Constraint (Pair left _)) = findClassWorker left
-  where
-    -- TODO (thomas.tuegel): Don't use pattern synonyms here!
-    findClassWorker (Var_ _) = Variables
-    findClassWorker (ElemVar_ _) = Variables
-    findClassWorker (SetVar_ _) = Variables
-    findClassWorker (StringLiteral_ _) = ConcreteBuiltins
-    findClassWorker (InternalBytes_ _ _) = ConcreteBuiltins
-    findClassWorker (InternalInt_ _) = ConcreteBuiltins
-    findClassWorker (InternalBool_ _) = ConcreteBuiltins
-    findClassWorker (InternalString_ _) = ConcreteBuiltins
-    findClassWorker (App_ symbol _) =
-        if Symbol.isConstructor symbol
-            then ConstructorAtTop
-            else OtherTermLike
-    findClassWorker (InternalList_ _) = ListBuiltin
-    findClassWorker (InternalSet_ _) = AssocCommBuiltin
-    findClassWorker (InternalMap_ _) = AssocCommBuiltin
-    findClassWorker _ = OtherTermLike
-
-instance Ord variable => Ord (Constraint variable) where
-    compare constraint1@(Constraint pair1) constraint2@(Constraint pair2) =
-        compare (findClass constraint1) (findClass constraint2)
-            <> compare pair1 pair2
-
-{- | Dispatch a single matching constraint.
-
-@matchOne@ is the heart of the matching algorithm. Each matcher is applied to
-the constraint in sequence, until one accepts the pair. The matchers may
-introduce substitutions and new constraints. If none of the matchers accepts the
-pair, it is deferred until we have more information.
--}
-matchOne ::
-    Pair (TermLike RewritingVariableName) ->
-    MatcherT RewritingVariableName Simplifier ()
-matchOne pair =
-    ( matchVariable pair
-        <|> matchEqualHeads pair
-        <|> matchAnd pair
-        <|> matchExists pair
-        <|> matchForall pair
-        <|> matchApplication pair
-        <|> matchBuiltinList pair
-        <|> matchBuiltinMap pair
-        <|> matchBuiltinSet pair
-        <|> matchInj pair
-        <|> matchOverload pair
-    )
-        & Error.maybeT (defer pair) return
-
-{- | Drive @matchOne@ until it cannot continue.
-
-Matching ends when all constraints have been dispatched. If there are remaining
-deferred constraints, then matching fails.
--}
-matchIncremental ::
-    SideCondition RewritingVariableName ->
+type NormalizedAc normalized =
+    Builtin.NormalizedAc normalized Key (TermLike RewritingVariableName)
+
+needs ::
     TermLike RewritingVariableName ->
     TermLike RewritingVariableName ->
-    Simplifier (Maybe (MatchResult RewritingVariableName))
-matchIncremental sideCondition termLike1 termLike2 =
-    evalRWST matcher sideCondition initial
-        & fmap fst
+    Bool
+needs (InternalSet_ s) term = needsAc (builtinAcChild s) term
+needs (InternalMap_ m) term = needsAc (builtinAcChild m) term
+needs _ _ = False
```

Comment:
what if the InternalSet/Map is nested inside the pattern? Or does such case not arise/matter?

Reply:
This is covered by the false case; in such cases, the ordering of elements doesn't really matter and thus we don't need to do anything special.